### PR TITLE
chore(auth-server): enable typescript-eslint

### PR DESCRIPTION
**Pull request by julianpoy**
## Because

* We aren't currently linting TS files in auth-server

## This pull request

* Enables and fixes linting issues in auth-server

## Issue that this pull request solves

Closes FXA-6136

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [x] If applicable, I have modified or added tests which pass locally.
- [ ] I have added necessary documentation (if appropriate).
- [ ] I have verified that my changes render correctly in RTL (if appropriate).

### CODE REVIEW

File: packages/fxa-auth-server/lib/types.ts
Comment (biancadanforth) [inline packages/fxa-auth-server/lib/types.ts:81]: Hm... I guess one `AuthLogger` (and `ProfileClient`) is an `interface` which is a TS-only thing that gets removed when TS is compiled to JS, so in the JS, we will only have one variable here. Are you avoiding renaming any of them, since both the interface and types are exported and used elsewhere, requiring a large number of variable name changes throughout the source code?

File: packages/fxa-auth-server/scripts/paypal-processor.ts
Comment (biancadanforth) [inline packages/fxa-auth-server/scripts/paypal-processor.ts:108]: Not part of this PR, but I'm curious why we have this `for await..of` loop at all? Perhaps @chenba might be able to provide some insight as the author?

File: packages/fxa-auth-server/test/local/senders/renderer.ts
Comment (biancadanforth) [inline packages/fxa-auth-server/test/local/senders/renderer.ts:57]: Is this testing the same thing anymore? Might want to put an eslint ignore next line here?

File: packages/fxa-auth-server/test/bench/index.js
Comment (biancadanforth) [inline packages/fxa-auth-server/test/bench/index.js:16]: According to my local checkout of this branch, we don't need this eslint ignore line?

**Review by biancadanforth — COMMENTED** [review]
Hi Julian; I took a first pass pre-CircleCI config and have some comments for you. The main change request so far is to update the auth server README, since I noted when we paired that it was incorrect even for the previous state of linting.

File: packages/fxa-auth-server/test/local/senders/renderer.ts
Comment (julianpoy) [reply packages/fxa-auth-server/test/local/senders/renderer.ts:57]: The tests before/after are equivalent to one-another.

File: packages/fxa-auth-server/lib/types.ts
Comment (julianpoy) [reply packages/fxa-auth-server/lib/types.ts:81]: Yes - avoiding renaming. I went down the path of attempting a rename and I decided to leave that as out of scope. Our typenames should generally not be the same as variable names in the future. IProfileClient/IAuthLogger is always a good way to handle this.

File: packages/fxa-auth-server/lib/types.ts
Comment (julianpoy) [reply packages/fxa-auth-server/lib/types.ts:47]: If it's the other way around, 3 ignores are needed.

File: packages/fxa-auth-server/lib/senders/renderer/index.ts
Comment (julianpoy) [reply packages/fxa-auth-server/lib/senders/renderer/index.ts:133]: Because of the directly preceeding comment.

File: package.json
Comment (julianpoy) [reply package.json:57]: Correct - eslint looks recursively upwards for an .eslintrc file from the given file to be linted.

File: _dev/.eslintrc
Comment (julianpoy) [reply _dev/.eslintrc]: Outdated file

File: _dev/.eslintrc
Comment (julianpoy) [reply _dev/.eslintrc]: Outdated file. You must have loaded before the push where I removed the yarn workspaces approach when the PR was a draft.

File: _dev/.eslintrc
Comment (biancadanforth) [reply _dev/.eslintrc]: Okay thanks. This question still applies to `packages/fxa-auth-server/.eslintrc.json` rule changes though.

Comment (biancadanforth) [reply _dev/.eslintrc]: Also, AFAICT, we're not using `_dev/.eslintrc` anywhere -- so can we remove that file entirely now? It's not going to get used automatically based on https://eslint.org/docs/latest/user-guide/configuring/configuration-files#cascading-and-hierarchy, since `_dev` is not a parent directory of any of the files being linted.

File: package.json
Comment (biancadanforth) [reply package.json:64]: Want to raise this at our next team meeting? Perhaps this is something our test engineering teammates can help with?

File: packages/fxa-auth-server/lib/senders/renderer/index.ts
Comment (biancadanforth) [reply packages/fxa-auth-server/lib/senders/renderer/index.ts:133]: How does that comment relate? It doesn't mention why we need a `catch` block -- only why we need a `require`.

File: packages/fxa-auth-server/scripts/paypal-processor.ts
Comment (chenba) [reply packages/fxa-auth-server/scripts/paypal-processor.ts:108]: It's so that processor.processInvoices() yields until it's done.

processor.processInvoices() was written in that way because we want to check the lock between each yield; see line 95.  Here there's no lock so we just have an empty block for the for..of.

File: packages/fxa-auth-server/test/local/senders/renderer.ts
Comment (biancadanforth) [reply packages/fxa-auth-server/test/local/senders/renderer.ts:57]: I understand that the escape is not needed in the template string, but I'm referring to what the test is trying to check.

IIUC, the test is checking to see that even if escape characters are used in the email text, that the method correctly splits the key from the value when we're trying to localize the text. If we remove those escape chars from the input, then this is no longer testing what it intended to test, right?

File: .circleci/config.yml
Comment (biancadanforth) [inline .circleci/config.yml:144]: I see that this workflow runs in parallel with our other workflows, so at the end of the day, it's not increasing the wall clock time of our CI overall, but it did take nearly 3 minutes to run (with more than 90% of that time setting up and tearing down). Compare this to less than 6 seconds for `yarn lint` to run locally (for me anyway).

What I'm trying to say is: I'm wondering if it makes sense to only add this to the Husky pre-commit hook and leave it out of CI? Maybe it's worth checking in with @clouserw or others on the team to see what they think (e.g. asking in #fxa-team)?

**Review by biancadanforth — COMMENTED** [review]
This is looking pretty nice! The main things I'm looking for are:
1. Are we sure we want to create a new `lint` workflow over say, only using a pre-commit hook? (ideally seek some input from the team)
2. auth server README update on the updated `yarn lint` command
There are a couple of other minor loose end types of questions outstanding from my first pass. Maybe we can discuss all of this sync in our pairing session tomorrow morning?

**Comment by pdehaan** [comment]
Also verified there weren't any other potential TypeScript-lintable files (like *.jsx or *.tsx) in fxa-auth-server folder:
```sh
git ls-files packages/fxa-auth-server | grep -Ev "\.(js|ts|ftl|sql|txt|json|sh|ini|mjml|lua|md|scss)"

packages/fxa-auth-server/.nsprc
packages/fxa-auth-server/.prettierignore
packages/fxa-auth-server/AUTHORS
packages/fxa-auth-server/LICENSE
packages/fxa-auth-server/docs/onepw.graffle
packages/fxa-auth-server/docs/self-host.docker
packages/fxa-auth-server/lib/senders/emails/storybook.css
packages/fxa-auth-server/scripts/.eslintrc
packages/fxa-auth-server/scripts/api-docs.handlebars
packages/fxa-auth-server/test/.eslintrc
packages/fxa-auth-server/test/load/Makefile
packages/fxa-auth-server/test/load/loadtests.py
```

As well as verified that *.ts files were being linted via:

```sh
DEBUG=eslint:cli-engine yarn lint
  eslint:cli-engine Lint /Volumes/Dev/github/mozilla/fxa/packages/fxa-auth-server/Gruntfile.js +0ms
  eslint:cli-engine Lint /Volumes/Dev/github/mozilla/fxa/packages/fxa-auth-server/bin/email_notifications.js +27ms
  eslint:cli-engine Lint /Volumes/Dev/github/mozilla/fxa/packages/fxa-auth-server/bin/key_server.js +21ms
  eslint:cli-engine Lint /Volumes/Dev/github/mozilla/fxa/packages/fxa-auth-server/bin/profile_server_messaging.js +26ms
  eslint:cli-engine Lint /Volumes/Dev/github/mozilla/fxa/packages/fxa-auth-server/bin/update_firestore_from_oauth.js +4ms
  eslint:cli-engine Lint /Volumes/Dev/github/mozilla/fxa/packages/fxa-auth-server/config/index.ts +9ms
  ...
```

File: packages/fxa-auth-server/scripts/paypal-processor.ts
Comment (pdehaan) [inline packages/fxa-auth-server/scripts/paypal-processor.ts:93]: We could potentially also use something like https://eslint.org/docs/latest/rules/no-unused-vars#varsignorepattern in the base .eslintrc config with some regex like `/^_$/` to have it ignore `_` variable names (although possibly a gross over-complication if we're only using this in 2-3 places).

File: packages/fxa-auth-server/config/index.ts
Comment (pdehaan) [inline packages/fxa-auth-server/config/index.ts:18]: Q: Why the rename, was ESLint complaining about `conf`?

File: .circleci/config.yml
Comment (pdehaan) [inline .circleci/config.yml:142]: Should we use something like this to see if we can do the CI linting in parallel?
```sh
yarn workspaces foreach --include=fxa-auth-server --include=fxa-shared --parallel run lint
# OR
yarn workspaces foreach --include="{fxa-auth-server,fxa-shared}" --parallel run lint
```

File: packages/fxa-auth-server/test/bench/index.js
Comment (biancadanforth) [reply packages/fxa-auth-server/test/bench/index.js:16]: It's not needed -- see https://github.com/mozilla/fxa/pull/14391#discussion_r1013195825.

File: _dev/.eslintrc
Comment (julianpoy) [reply _dev/.eslintrc]: Will happen in 6137, yep - already have that implemented in my local branch. I don't necessarily want to move this to a JS file.

Comment (biancadanforth) [reply _dev/.eslintrc]: Added a comment to https://mozilla-hub.atlassian.net/browse/FXA-6137?focusedCommentId=609569 to take care of this.

File: package.json
Comment (julianpoy) [reply package.json:57]: Would need to be moved up 1 directory so that all packages technically inherit from this base directory.

Comment (biancadanforth) [reply package.json:57]: Discussed sync with Julian; he's going to move the file up a directory, so that packages that don't have any .eslintrc files of their own still get linted according to this config.

File: package.json
Comment (biancadanforth) [reply package.json:64]: This is a longstanding issue. Even without this PR, there are JS files with linting errors that would be found by the Husky hook assuming everyone is using it and not bypassing it. Julian will notify the team about the potential here for linting errors to be surfaced on staged files in general.
Julian may open a follow up PR for any other packages that are close to passing eslint (i.e. not thousands of errors).

File: packages/fxa-auth-server/lib/senders/renderer/index.ts
Comment (biancadanforth) [reply packages/fxa-auth-server/lib/senders/renderer/index.ts:133]: I did notice it was a pretty critical path (the Mailer.prototype.send function), so I agree it's a little risky to change without digging a little deeper.

Comment (biancadanforth) [reply packages/fxa-auth-server/lib/senders/renderer/index.ts:133]: @xlisachan : Git blame suggests this was added by you in https://github.com/mozilla/fxa/commit/f2a175ae86906f588cf7228fd556dbab824eb4b9 -- do we need this `catch` block? What does it do for us?

File: .circleci/config.yml
Comment (julianpoy) [reply .circleci/config.yml:144]: > I'm wondering if it makes sense to only add this to the Husky pre-commit hook

This is how we get into the same situation we're in now - if someone doesn't use husky locally we'll have lint errors all over the place. We need CI acting as the source of truth here, and preventing us from merging code that violates our code standards.

In other words, this is definitely worth our CI time.

File: packages/fxa-auth-server/test/local/senders/renderer.ts
Comment (julianpoy) [reply packages/fxa-auth-server/test/local/senders/renderer.ts:57]: Those escape characters aren't even present in the string, then. If you wanted the escape character to be present in the string, they'd need to be `\\`.

Removing the test would not be part of this scope of work imo.

Comment (julianpoy) [reply packages/fxa-auth-server/test/local/senders/renderer.ts:57]: I just updated the test to use `\\` in local, and the test fails. This may be a real bug if that must be supported. We'll want to look into this later.

File: packages/fxa-auth-server/test/bench/index.js
Comment (julianpoy) [reply packages/fxa-auth-server/test/bench/index.js:16]: Pass on line 44 does not use the variable. It simply increments it. The linter is correct here, that this is an unused variable (it's value is never read, just set/updated).

Comment (biancadanforth) [reply packages/fxa-auth-server/test/bench/index.js:16]: Okay we tried this locally, and while only one eslint rule is needed, eslint still marks them both as errors in the IDE. We'll just leave them both there as they're innocuous.

File: .circleci/config.yml
Comment (julianpoy) [reply .circleci/config.yml:142]: Since this destroys any formatting/coloring and makes it harder to parse the output here, I'd like to hold off on this. At the moment lint takes only 18 seconds of job time.

Would be happy to re-address this later if we find that this starts to grow to more than 2/3 minutes.

File: packages/fxa-auth-server/.eslintrc.json
Comment (julianpoy) [reply packages/fxa-auth-server/.eslintrc.json]: Done, thanks!

**Review by biancadanforth — APPROVED** [review]
R+ (assuming CI passes and things are squashed). Thanks Julian; I believe we've resolved all of the open questions I had during my review sync.

As we discussed, it would be good to let the team know about this change and what they can expect going forward before merging.

File: packages/fxa-auth-server/test/local/senders/renderer.ts
Comment (biancadanforth) [reply packages/fxa-auth-server/test/local/senders/renderer.ts:57]: Thanks for looking into it a bit more -- can you file a follow-up for this and link this comment?

**Pull request merged**
